src/python/pants/backend/python/rules/pex_test.py

```diff
@@ -26,6 +28,63 @@
 from pants.util.strutil import create_path_env_var
 
 
+def test_merge_interpreter_constraints() -> None:
+    def assert_merged(*, input: List[List[str]], expected: List[str]) -> None:
+        assert PexInterpreterConstraints.merge_constraint_sets(input) == expected
+
+    # Multiple constraint sets get merged so that they are ANDed.
+    # A & B => A & B
+    assert_merged(
+        input=[["CPython==2.7.*"], ["CPython==3.6.*"]], expected=["CPython==2.7.*,==3.6.*"]
+    )
+
+    # Multiple constraints within a single constraint set are kept separate so that they are ORed.
+    # A | B => A | B
+    assert_merged(
+        input=[["CPython==2.7.*", "CPython==3.6.*"]], expected=["CPython==2.7.*", "CPython==3.6.*"]
+    )
+
+    # Input constraints already were ANDed.
+    # A => A
+    assert_merged(input=[["CPython>=2.7,<3"]], expected=["CPython>=2.7,<3"])
+
+    # Both AND and OR.
+    # (A | B) & C => (A & B) | (B & C)
+    assert_merged(
+        input=[["CPython>=2.7,<3", "CPython>=3.5"], ["CPython==3.6.*"]],
+        expected=["CPython>=2.7,<3,==3.6.*", "CPython>=3.5,==3.6.*"],
+    )
+    # A & B & (C | D) => (A & B & C) | (A & B & D)
+    assert_merged(
+        input=[["CPython==2.7.*"], ["CPython==3.6.*"], ["CPython==3.7.*", "CPython==3.8.*"]],
+        expected=["CPython==2.7.*,==3.6.*,==3.7.*", "CPython==2.7.*,==3.6.*,==3.8.*"],
+    )
+    # (A | B) & (C | D) => (A & C) | (A & D) | (B & C) | (B & D)
```

Comment:
Can we add one really crazy test case, for A & (B | C | D) & (E | F) & G, just to be sure we can handle it?  Should yield

(A & B & E & G) | (A & B & F & G) | (A & C & E & G) | (A & C & F & G) | (A & D & E & G) | (A & D & F & G)

Reply:
Sure, great idea!